### ISSUE TITLE
Resurrect Tensorflow Lite library

### ISSUE THREAD

**Issue by Shillaker** [faasm/faasm#651]
We've not run the TFLite code for several years now, but it should be possible to resurrect it.

The code lives in this archived repository: https://github.com/faasm/faasm-experiments.

The script to build the Tensorflow Lite library to WebAssembly is here: https://github.com/faasm/faasm-experiments/blob/master/tasks/tensorflow.py#L38

While the function to invoke it is here: https://github.com/faasm/faasm-experiments/tree/master/func/tf

The steps to recreate the experiment results with Faasm now would be:

- Add Tensorflow as a submodule in the Faasm cpp repo (https://github.com/faasm/cpp)
- Create a task to build TFLite similar to how we build the ffmpeg library
- Move the function code from faasm-experiments into the cpp repo too (i.e. under https://github.com/faasm/cpp/tree/main/func/tf)

**Comment by FlowDrucka64** [comment]
Hello!
I made good progress in ressurecting the TF lib so far.

Compillation and installation of the libary seems to work.
```
(cpp) root@ff21b88627ac:/code/cpp# inv tensorflow.lib
make: Entering directory '/code/cpp/third-party/tensorflow'
make: Nothing to be done for 'lib'.
make: Leaving directory '/code/cpp/third-party/tensorflow'
```

Compillation of the invoke function seems to work.
```
(cpp) root@ff21b88627ac:/code/cpp# inv tensorflow.func
cmake -GNinja -DFAASM_BUILD_TYPE=wasm -DCMAKE_TOOLCHAIN_FILE=/usr/local/faasm/toolchain/tools/WasiToolchain.cmake -DCMAKE_BUILD_TYPE=Release /code/cpp/func
-- Faasm building STATIC libraries
-- Faasm building target wasm32-wasi
System is unknown to cmake, create:
Platform/Wasm to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
Your CMakeCache.txt file was copied to CopyOfCMakeCache.txt. Please post that file on discourse.cmake.org.
-- Detected wasm build (sysroot=/usr/local/faasm/llvm-sysroot)
-- Configuring done
-- Generating done
-- Build files have been written to: /code/cpp/build/func
ninja: no work to do.
```

Upload of the state seems to work.
```
(cpp) root@ff21b88627ac:/code/cpp# inv tensorflow.state
Uploading binary file at /code/cpp/func/tf/data/mobilenet_v1_1.0_224.tflite for user tf
curl -X PUT http://upload:8002/s/tf/mobilenet_v1 -T /code/cpp/func/tf/data/mobilenet_v1_1.0_224.tflite
State upload complete
Successfully PUT file /code/cpp/func/tf/data/mobilenet_v1_1.0_224.tflite to http://upload:8002/s/tf/mobilenet_v1
```

Upload of the data seems to work.
```
(cpp) root@ff21b88627ac:/code/cpp# inv tensorflow.upload
Uploading mobilenet_v1_1.0_224.tflite to tfdata/mobilenet_v1_1.0_224.tflite
curl -X PUT http://upload:8002/file/ -T /code/cpp/func/tf/data/mobilenet_v1_1.0_224.tflite -H "FilePath: tfdata/mobilenet_v1_1.0_224.tflite"
Shared file uploaded
Successfully PUT file /code/cpp/func/tf/data/mobilenet_v1_1.0_224.tflite to http://upload:8002/file/
Uploading grace_hopper.bmp to tfdata/grace_hopper.bmp
curl -X PUT http://upload:8002/file/ -T /code/cpp/func/tf/data/grace_hopper.bmp -H "FilePath: tfdata/grace_hopper.bmp"
Shared file uploaded
Successfully PUT file /code/cpp/func/tf/data/grace_hopper.bmp to http://upload:8002/file/
Uploading labels.txt to tfdata/labels.txt
curl -X PUT http://upload:8002/file/ -T /code/cpp/func/tf/data/labels.txt -H "FilePath: tfdata/labels.txt"
Shared file uploaded
Successfully PUT file /code/cpp/func/tf/data/labels.txt to http://upload:8002/file/
```

Upload of the function seems to work:
```
(cpp) root@ff21b88627ac:/code/cpp# inv func.upload tf image
Response 200: Function upload complete
```

However when trying to invoke the function:
```
(cpp) root@ff21b88627ac:/code/cpp# inv func.flush func.invoke tf image
Flush response 200: Flush sent
Error (500):
Call failed (return value=1)
```

I will start trying to figure out why its not working soon but I thought I'd share it and see if you have any suspicions beforehand.

**Comment by FlowDrucka64** [comment]
Came back to it today and made it work!
The issue was, that the main function of the "old" function code was called ``FAASM_MAIN_FUNC()``. Changed it to ``int main(int argc, char* argv[])`` and it works like a charm!

**Comment by csegarragonz** [comment]
Hey @FlowDrucka64, great news that you made it work!

Any chance that you could upstream the working version to `faasm/cpp`?

Ideally we'd also two tests:
* One in `faasm/cpp` to check that compilation works succesfully ([example of FFmpeg test](https://github.com/faasm/cpp/blob/main/.github/workflows/tests.yml#L51)).
* One in `faasm/faasm` to check that a basic function actuall runs. The test function would be in `tests/test/ported_libs` together with the FFmpeg one.

Thank you very much for your work in resurrecting TF!

**Comment by FlowDrucka64** [comment]
Yeah sure I'd like to do that!
I'm not certain how to do that exactly though since my changes are currently not reflected inside your cpp image.
Do I create a new image with my changes and run the tests for this image?
Also I think I dont want to upstream until I've finished my work on this (currently working on ressurecting the TF benchmarking) since there might be further changes.
By the way I forked the [cpp ](https://github.com/FlowDrucka64/cpp) repo so you can have a look at the changes I made if you want.

**Comment by csegarragonz** [comment]
Great! I have opened a PR in `faasm/cpp` with your changes, and will add comments there: faasm/cpp#96

**Comment by csegarragonz** [comment]
Closed with faasm/examples#1
